src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs

```diff
@@ -91,16 +92,15 @@ public async ValueTask DisposeAsync()
             if (!_disposed)
             {
                 _disposed = true;
-                await _stream.DisposeAsync().ConfigureAwait(false);
-                DisposeSyncHelper();
+                var stream = Interlocked.Exchange(ref _stream, null!);
+                await stream.DisposeAsync().ConfigureAwait(false);
+                DisposeSyncHelper(stream);
             }
         }
 
-        private void DisposeSyncHelper()
+        private void DisposeSyncHelper(QuicStream stream)
         {
-            _connection.RemoveStream(_stream);
-            _connection = null!;
-            _stream = null!;
+            Interlocked.Exchange(ref _connection, null!).RemoveStream(stream);
```

Comment:
If we've got sole ownership of the stream now due to the Interlocked in the dispose methods above, do we still need this interlocked?  Or is there another caller that could get here outside of Dispose{Async}?

Reply:
I've mostly done that to be sure the change is visible to null checks in `Http3RequestStream.HandleReadResponseContentException` https://github.com/dotnet/runtime/pull/54334/files#diff-b79affd636899ac98b816c1398f9dca19bc4850a939a3815c09c5a9931094a4cR1113-R1120 In the issue I'm trying to solve, disposal and this exception handling are happening concurrently. Here https://github.com/dotnet/runtime/pull/52800#discussion_r647135975 @scalablecory says the change might not be visible otherwise. On the other hand, @wfurt offline told me that we usually don't worry about this for example for setting and checking `_disposed` flag. I am not sure what should be the best practice here.